pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,665 @@
+
+		// update CRI
+		if err := r.update(ctx, azVolumeAttachment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
+		return err
+	}
+
+	// TODO reassess if adding additional finalizer to AzVolume CRI is necessary
+	// // add finalizer to the bound AzVolume CRI
+	// conditionFunc = func() (bool, error) {
+	// 	if err := r.addFinalizerToAzVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume); err != nil {
+	// 		return false, nil
+	// 	}
+	// 	return true, nil
+	// }
+	if err := wait.PollImmediate(updateAttemptInterval, updateTimeout, conditionFunc); err != nil {
+		return err
+	}
+
+	// initiate goroutine to attach volume
+	go func() {
+		response, err := r.attachVolume(ctx, azVolumeAttachment.Spec.VolumeID, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Spec.VolumeContext)
+		var conditionFunc func() (bool, error)
+		if err != nil {
+			klog.Errorf("failed to attach volume %s to node %s: %v", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, err)
+
+			conditionFunc = func() (bool, error) {
+				azVolumeAttachment, derr := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, false)
+				if errors.IsNotFound(derr) {
+					// if failed to get azVolumeAttachment then escape the wait
+					return false, derr
+				} else if derr != nil {
+					klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, derr)
+					return false, nil
+				}
+				azVolumeAttachment = azVolumeAttachment.DeepCopy()
+				azVolumeAttachment = r.updateState(ctx, azVolumeAttachment, v1alpha1.AttachmentFailed)
+				azVolumeAttachment = r.updateError(ctx, azVolumeAttachment, err)
+				if derr := r.update(ctx, azVolumeAttachment); derr != nil {
+					return false, nil
+				}
+				return true, nil
+			}
+		} else {
+			klog.Infof("successfully attached volume (%s) to node (%s) and update status of AzVolumeAttachment (%s)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, azVolumeAttachment.Name)
+
+			conditionFunc = func() (bool, error) {
+				azVolumeAttachment, derr := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, false)
+				if derr != nil {
+					klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, derr)
+					// if failed to get azVolumeAttachment then escape the wait
+					return false, derr
+				}
+				azVolumeAttachment = azVolumeAttachment.DeepCopy()
+				azVolumeAttachment = r.updateState(ctx, azVolumeAttachment, v1alpha1.Attached)
+				azVolumeAttachment = r.updateStatusDetail(ctx, azVolumeAttachment, response)
+				if derr := r.update(ctx, azVolumeAttachment); derr != nil {
+					return false, nil
+				}
+				return true, nil
+			}
+		}
+		if derr := wait.PollImmediateInfinite(updateAttemptInterval, conditionFunc); derr != nil {
```

Comment:
I am assuming the purpose of this polling-wait is to retry the `azVolumeAttachment` update on a conflict. Is that correct? A comment would be helpful.

Reply:
Yes, that's correct. I will add comments to clarify things :)

Reply:
It'd be nice to have a common function to handle this and make the purpose of the loop clear, e.g. `func (r *ReconcileAttachDetach) updateWithRetryOnConflict(azVolumeAttachment *v1alpha1.AzVolumeAttachment, updateFunc func (*v1alpha1.AzVolumeAttachment) err)`.